### ISSUE TITLE
Start Approval Workflow Action Fails (First Time Only) From an Alfresco Transaction Issue 

### ISSUE THREAD

**Issue by mwojno-tsg** [tsgrp/HPI#2310]
## Overview
- While testing the 3.4.2 release in a test environment before releasing officially, we noticed the Start Approval Workflow action failed on the first attempt. Each time after that the action was run it succeeded and workflow was able to be successfully completed. 

- The issue that was happening was the `approval_details.json` was trying to be created twice with the same object name which is not possible. Here is the order that the start approval workflow will follow for creating/updating the approval details in `PrepareWizardApprovalServiceTask`:
1. `resetAndMaintainApprovalDetails` will create a new `approval_details.json` in the folder `Wizard/wizard/formName/` by calling `setApprovalDetails` within the function. 
2. `setApprovalDetails` called next _should_ find the already created `approval_details.json` from a `getChildren` call. However, the transaction had not been committed so it thinks it needs to create a new `approval_details.json`. This is where the error occurs b/c it already has been created.

## Steps to Reproduce
- Create a Change Request form. Start the approval workflow using the action. 
- It will fail almost immediately with a 500 error. 
- Start the approval workflow using the action again. Every time after it will succeed because the `approval_details.json` exists and can be queried. 

## Insights 
- This was a particularly challenging issue because we were unable to reproduce the error locally. We could only see this happen within the test environment setup by the deployment team. 

- Transactions were the culprit here because the `PrepareWizardApprovalServiceTask` opened a parent transaction. Then within the service task the act of creating the approval details were queued up after the parent transaction closed. 

- We either suspect this to be an eventual consistency and or transaction order issue. 

- The eventual consistency makes sense because it happens almost one after another that one function calls to create the `approval_details.json` while the next function tries to query for the newly created `approval_details.json` . So, it could be a timing issue where we are querying too quickly for the database to sync with the application.

- The transaction order also could make sense because the `doEnhancedCreateObject` forces a new transaction to be opened. This new transaction could be queued for after the parent transaction closes which could make sense that two creation could be queued causing the name issue.

## Potential Solution
- A solution that we saw finally working was that we would not require a new transaction to be opened for the `doEnhancedUpdateObject` or the `doEnhancedCreateObject`. This would let both actions happen within the wrapped parent transaction. 

- We coupled the above changes with an update to the `setApprovalDetails`  and `resetAndMaintainApprovalDetails` function signature to return a `String` , which would be the approval details object id. Also, the `setApprovalDetails` would have an optional parameter of the approval details object id. This way we could return the approval object id from the first call, `resetAndMaintainApprovalDetails` and pass it to the second call for `setApporvalDetails`.

- This would let us add logic ^ to the `setApprovalDetails` to not perform a query for the `approval_details.json` if the object id is already passed into the function. 

- A final change we made that seemed to work was to add the `setApprovalDetails` to the restricted list for the `EmbeddedAlfrescoAdvice` to not wrap in a transaction.

**Comment by gsteimer** [comment]
@mwojno-tsg - to be clear, this only happens on very specific environments, correct?  We haven't seen this in any of our test environments.

**Comment by mwojno-tsg** [comment]
@gsteimer I tested almost all versions of alfresco 6 with 3.4.2 pnp release code for the happy path and didn't see this issue. 3.4.1 had no issues. Customer environment on 3.4.1 had no issues. Code around this Start Approval Workflow should not have changed from 3.4.1 -> 3.4.2.

**Comment by mwojno-tsg** [comment]
This was closed with OC commit: **26378**. The following solution was followed above. We changed the function signature on `resetAndMaintainApprovalDetails` to return a string aka the approval details id. To then be passed into the `setApprovalDetails` as an optional parameter within `PrepareWizardApprovalServiceTask`. Everywhere else for `setApprovalDetails` passes in null and runs the same `getChildren` it had already done previously. 

CR: aking, dgrumieaux
